numba_scipy/sparse.py

```diff
@@ -0,0 +1,225 @@
+import warnings
+import numpy as np
+import scipy as sp
+import scipy.sparse
+from numba import types
+
+try:
+    from numba.core import cgutils
+    from numba.core.imputils import impl_ret_borrowed
+except ImportError:
+    from numba import cgutils
+    from numba.targets.imputils import impl_ret_borrowed
+
+try:
+    from numba.core.errors import NumbaExperimentalFeatureWarning
+except ImportError:
+    from numba.errors import NumbaWarning
+
+    class NumbaExperimentalFeatureWarning(NumbaWarning):
+        pass
+
+
+from numba.extending import (
+    NativeValue,
+    box,
+    intrinsic,
+    make_attribute_wrapper,
+    models,
+    overload,
+    overload_attribute,
+    overload_method,
+    register_model,
+    typeof_impl,
+    unbox,
+)
+
+
+class CSMatrixType(types.Type):
+    """A Numba `Type` modeled after the base class `scipy.sparse.compressed._cs_matrix`."""
+
+    name: str
+
+    @staticmethod
+    def instance_class(data, indices, indptr, shape):
+        raise NotImplementedError()
+
+    def __init__(self, dtype):
+        warnings.warn(
+            "Use of sparse matrices detected. This is an experimental feature.",
+            category=NumbaExperimentalFeatureWarning,
+        )
+        self.dtype = dtype
+        self.data = types.Array(dtype, 1, "A")
+        self.indices = types.Array(types.int32, 1, "A")
+        self.indptr = types.Array(types.int32, 1, "A")
```

Comment:
Minor thing, but these can also be `int64`. Scipy is pretty aggressive about down casting it to int32, but if you have enough values to need 64 bit integers they will stick around.

Reply:
> Minor thing, but these can also be `int64`. Scipy is pretty aggressive about down casting it to int32, but if you have enough values to need 64 bit integers they will stick around.
> 
> Super excited to see this moving forward!

Thanks for the heads up!  If no one objects, I'll be glad to change them.

Reply:
Linking in some relevant discussion from scipy, since it came up on my notifications again: https://github.com/scipy/scipy/issues/16774

tldr (from my perspective) scipy wants to be more supportive of int64 indices, and remove the automatic down casting to int32 since checking the values is expensive.